cloud/linode/firewall/firewalls.go

```diff
@@ -0,0 +1,463 @@
+	devices, err := l.Client.ListFirewallDevices(ctx, firewalls[0].ID, &linodego.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if len(devices) == 0 {
+		// nothing attached to it, clean it up
+		return l.Client.DeleteFirewall(ctx, firewalls[0].ID)
+	}
+	// else let that firewall linger, don't mess with it.
+
+	return nil
+}
+
+// getNodeBalancerDeviceID gets the deviceID of the nodeBalancer that is attached to the firewall.
+func (l *LinodeClient) getNodeBalancerDeviceID(ctx context.Context, firewallID, nbID int) (int, bool, error) {
+	devices, err := l.Client.ListFirewallDevices(ctx, firewallID, &linodego.ListOptions{})
+	if err != nil {
+		return 0, false, err
+	}
+
+	if len(devices) == 0 {
+		return 0, false, nil
+	}
+
+	for _, device := range devices {
+		if device.Entity.ID == nbID {
+			return device.ID, true, nil
+		}
+	}
+
+	return 0, false, nil
+}
+
+// Updates a service that has a firewallID annotation set.
+// If an annotation is set, and the nodebalancer has a firewall that matches the ID, nothing to do
+// If there's more than one firewall attached to the node-balancer, an error is returned as its not a supported use case.
+// If there's only one firewall attached and it doesn't match what's in the annotation, the new firewall is attached and the old one removed
+func (l *LinodeClient) updateServiceFirewall(ctx context.Context, service *v1.Service, nb *linodego.NodeBalancer) error {
+	var newFirewallID int
+	var err error
+
+	fwID := service.GetAnnotations()[annotations.AnnLinodeCloudFirewallID]
+	newFirewallID, err = strconv.Atoi(fwID)
+	if err != nil {
+		return err
+	}
+
+	// See if a firewall is attached to the nodebalancer first.
+	firewalls, err := l.Client.ListNodeBalancerFirewalls(ctx, nb.ID, &linodego.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(firewalls) > 1 {
+		klog.Errorf("Found more than one firewall attached to nodebalancer: %d, firewall IDs: %v", nb.ID, firewalls)
+		return ErrTooManyNBFirewalls
+	}
+
+	// get the ID of the firewall that is already attached to the nodeBalancer, if we have one.
+	var existingFirewallID int
+	if len(firewalls) == 1 {
+		existingFirewallID = firewalls[0].ID
+	}
+
+	// if existing firewall and new firewall differs, attach the new firewall and remove the old.
+	if existingFirewallID != newFirewallID {
+		// attach new firewall.
+		_, err = l.Client.CreateFirewallDevice(ctx, newFirewallID, linodego.FirewallDeviceCreateOptions{
+			ID:   nb.ID,
+			Type: "nodebalancer",
+		})
+		if err != nil {
+			return err
+		}
+		// remove the existing firewall if it exists
+		if existingFirewallID != 0 {
+			deviceID, deviceExists, err := l.getNodeBalancerDeviceID(ctx, existingFirewallID, nb.ID)
+			if err != nil {
+				return err
+			}
+
+			if !deviceExists {
+				return fmt.Errorf("error in fetching attached nodeBalancer device")
+			}
+
+			err = l.Client.DeleteFirewallDevice(ctx, existingFirewallID, deviceID)
+			if err != nil {
```

Comment:
nit:
```suggestion
			if err := l.Client.DeleteFirewallDevice(ctx, existingFirewallID, deviceID); err != nil {
```

Reply:
updated and in a few other spots in this file too